File: RoutinUsage/Providers/GLMUsageProvider.swift
import Foundation

struct GLMUsageProvider: UsageProvider {
    let descriptor: ProviderDescriptor
    private let session: URLSession
    private let baseURL: URL

    init(session: URLSession = .shared, baseURL: URL = URL(string: "https://api.z.ai")!) {
        self.session = session
        self.baseURL = baseURL
        self.descriptor = ProviderRegistry.builtInDescriptors.first(where: { $0.id == .glm })!
    }

    func validate(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        try await fetchUsage(credential, now: now)
    }

    func fetchUsage(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        guard credential.providerID == .glm, credential.kind == .apiKey, !credential.secret.isEmpty else {
            throw UsageProviderError.invalidCredential
        }

        let root = credential.metadata["baseURL"]
            .flatMap(URL.init(string:)) ?? baseURL
        let start = Self.format(Self.windowStart(from: now))
        let end = Self.format(Self.windowEnd(from: now))
        let query = "?startTime=\(Self.encode(start))&endTime=\(Self.encode(end))"

        async let model = request(
            url: root.appendingPathComponent("/api/monitor/usage/model-usage").appending(query: query),
            credential: credential
        )
        async let quota = request(
            url: root.appendingPathComponent("/api/monitor/usage/quota/limit"),
            credential: credential
        )

        let (modelValue, quotaValue) = try await (model, quota)
        let quotaMetrics = Self.quotaMetrics(from: quotaValue)
        var metrics = quotaMetrics.filter { $0.id != "zcode-mcp" }
        if let modelCallMetric = Self.modelCallMetric(from: modelValue) {
            metrics.append(modelCallMetric)
        }
        if let zcodeMCPMetric = quotaMetrics.first(where: { $0.id == "zcode-mcp" }) {
            metrics.append(zcodeMCPMetric)
        }
        guard !metrics.isEmpty else {
            throw UsageProviderError.invalidResponse
        }

        return UsageSnapshot(
            planName: "Coding Plan",
            kind: .periodic,
            fiveHour: nil,
            weekly: nil,
            token: nil,
            allowedModels: [],
            fetchedAt: now,
            providerID: .glm,
            credentialID: credential.credentialID,
            metrics: metrics
        )
    }

    private func request(url: URL, credential: ProviderCredential) async throws -> GLMJSONValue {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        request.setValue(credential.secret, forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("en-US,en", forHTTPHeaderField: "Accept-Language")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw UsageProviderError.transport
        }
        guard let httpResponse = response as? HTTPURLResponse else {
            throw UsageProviderError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            switch httpResponse.statusCode {
            case 401, 403: throw UsageProviderError.unauthorized
            case 429: throw UsageProviderError.rateLimited
            case 500...599: throw UsageProviderError.providerUnavailable
            default: throw UsageProviderError.invalidResponse
            }
        }
        do {
            return try JSONDecoder().decode(GLMJSONValue.self, from: data)
        } catch {
            throw UsageProviderError.invalidResponse
        }
    }

    private static func quotaMetrics(from value: GLMJSONValue) -> [NormalizedUsageMetric] {
        let limits: [GLMJSONValue]
        switch value {
        case let .object(object):
            if case let .object(data)? = object["data"], case let .array(items)? = data["limits"] {
                limits = items
            } else if case let .array(items)? = object["limits"] {
                limits = items
            } else {
                limits = []
            }
        default:
            limits = []
        }

        var fiveHour: NormalizedUsageMetric?
        var weekly: NormalizedUsageMetric?
        var zcodeMCP: NormalizedUsageMetric?

        for item in limits {
            guard case let .object(object) = item,
                  let percentage = object["percentage"]?.numberValue()
            else { continue }
            let type = object["type"]?.stringValue() ?? "quota"
            let unit = object["unit"]?.numberValue().map { NSDecimalNumber(decimal: $0).intValue }
            let number = object["number"]?.numberValue().map { NSDecimalNumber(decimal: $0).intValue }
            let windowEnd = Self.date(fromMilliseconds: object["nextResetTime"]?.numberValue())

            switch (type, unit, number) {
            case ("TOKENS_LIMIT", 3, 5):
                fiveHour = Self.percentMetric(
                    id: "five-hour",
                    label: "5 小时用量",
                    percentage: percentage,
                    windowEnd: windowEnd
                )
            case ("TOKENS_LIMIT", 6, 1):
                weekly = Self.percentMetric(
                    id: "weekly",
                    label: "每周用量",
                    percentage: percentage,
                    windowEnd: windowEnd
                )
            case ("TIME_LIMIT", _, _):
                let limit = object["usage"]?.numberValue() ?? 100
                let used = object["currentValue"]?.numberValue() ?? percentage
                let remaining = object["remaining"]?.numberValue() ?? max(0, limit - used)
                zcodeMCP = NormalizedUsageMetric(
                    id: "zcode-mcp",
                    label: "ZCode MCP 用量",
                    used: used,
                    limit: limit,
                    remaining: remaining,
                    unit: .request,
                    windowEnd: windowEnd,
                    presentation: .progress,
                    healthState: Self.healthState(for: percentage)
                )
            default:
                continue
            }
        }

        return [fiveHour, weekly, zcodeMCP].compactMap { $0 }
    }

    private static func modelCallMetric(from value: GLMJSONValue) -> NormalizedUsageMetric? {
        guard case let .object(object) = value,
              case let .object(data)? = object["data"],
              case let .object(totalUsage)? = data["totalUsage"],
              let count = totalUsage["totalModelCallCount"]?.numberValue()
        else {
            return nil
        }
        return NormalizedUsageMetric(
            id: "model-calls",
            label: "调用量",
            value: count,
            unit: .request,
            presentation: .value,
            healthState: .normal
        )
    }

    private static func percentMetric(
        id: String,
        label: String,
        percentage: Decimal,
        windowEnd: Date?
    ) -> NormalizedUsageMetric {
        NormalizedUsageMetric(
            id: id,
            label: label,
            used: percentage,
            limit: 100,
            remaining: max(0, 100 - percentage),
            unit: .token,
            windowEnd: windowEnd,
            presentation: .progress,
            healthState: healthState(for: percentage)
        )
    }

    private static func healthState(for percentage: Decimal) -> UsageMetricHealthState {
        let value = NSDecimalNumber(decimal: percentage).doubleValue
        if value >= 80 { return .critical }
        if value >= 50 { return .warning }
        return .normal
    }

    private static func date(fromMilliseconds value: Decimal?) -> Date? {
        guard let value else { return nil }
        return Date(timeIntervalSince1970: NSDecimalNumber(decimal: value).doubleValue / 1_000)
    }

    private static func windowStart(from date: Date) -> Date {
        Calendar.current.date(byAdding: .day, value: -1, to: date) ?? date
    }

    private static func windowEnd(from date: Date) -> Date {
        Calendar.current.date(bySetting: .minute, value: 59, of: date)
            .flatMap { Calendar.current.date(bySetting: .second, value: 59, of: $0) } ?? date
    }

    private static func format(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return formatter.string(from: date)
    }

    private static func encode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
    }
}

private extension URL {
    func appending(query: String) -> URL {
        URL(string: absoluteString + query) ?? self
    }
}

File: android/provider/src/main/kotlin/ai/routin/mytoken/provider/glm/GLMUsageProvider.kt
package ai.routin.mytoken.provider.glm

import ai.routin.mytoken.domain.model.Credential
import ai.routin.mytoken.domain.model.CredentialKind
import ai.routin.mytoken.domain.model.CredentialMetadataKey
import ai.routin.mytoken.domain.model.CredentialSecret
import ai.routin.mytoken.domain.model.ProviderId
import ai.routin.mytoken.domain.model.UsageMetric
import ai.routin.mytoken.domain.model.UsageMetricHealthState
import ai.routin.mytoken.domain.model.UsageMetricPresentation
import ai.routin.mytoken.domain.model.UsageMetricSemantic
import ai.routin.mytoken.domain.model.UsageMetricUnit
import ai.routin.mytoken.domain.model.UsageSnapshot
import ai.routin.mytoken.domain.usage.UsageProvider
import ai.routin.mytoken.domain.usage.UsageProviderException
import ai.routin.mytoken.provider.health.UsageMetricHealthEvaluator
import ai.routin.mytoken.provider.http.HttpTransport
import ai.routin.mytoken.provider.http.ProviderHttpRequest
import ai.routin.mytoken.provider.http.ProviderHttpResponse
import ai.routin.mytoken.provider.json.ProviderJson
import ai.routin.mytoken.provider.json.arrayOrNull
import ai.routin.mytoken.provider.json.asObjectOrNull
import ai.routin.mytoken.provider.json.decimalOrNull
import ai.routin.mytoken.provider.json.intOrNull
import ai.routin.mytoken.provider.json.longOrNull
import ai.routin.mytoken.provider.json.objOrNull
import ai.routin.mytoken.provider.json.stringOrNull
import java.math.BigDecimal
import java.time.Clock
import java.time.Instant
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import kotlinx.serialization.json.JsonObject

/**
 * GLM Coding Plan usage adapter, aligned with the macOS GLMUsageProvider:
 * two GET requests (model usage + quota limit) with a raw Authorization header,
 * metric order [five-hour, weekly, model-calls, zcode-mcp].
 */
class GLMUsageProvider(
    private val transport: HttpTransport,
    private val clock: Clock = Clock.systemUTC(),
    private val defaultBaseURL: String = DEFAULT_BASE_URL,
) : UsageProvider {

    override val providerId: ProviderId = ProviderId.Glm

    override suspend fun fetchUsage(credential: Credential, secret: CredentialSecret): Result<UsageSnapshot> {
        if (credential.providerId != ProviderId.Glm || credential.credentialKind != CredentialKind.ApiKey) {
            return Result.failure(UsageProviderException.InvalidCredential())
        }
        val key = (secret as? CredentialSecret.ApiKey)?.key
            ?: return Result.failure(UsageProviderException.InvalidCredential())
        if (key.isEmpty()) return Result.failure(UsageProviderException.InvalidCredential())

        val baseURL = credential.metadata[CredentialMetadataKey.BaseURL]?.takeIf { it.isNotBlank() }
            ?: defaultBaseURL
        val query = "?startTime=${encode(windowStart())}&endTime=${encode(windowEnd())}"

        val modelResponse = try {
            request(url = baseURL.trimEnd('/') + MODEL_USAGE_PATH + query, key = key)
        } catch (error: Exception) {
            if (error is kotlinx.coroutines.CancellationException) throw error
            return Result.failure(error as? UsageProviderException ?: UsageProviderException.Transport())
        }
        val quotaResponse = try {
            request(url = baseURL.trimEnd('/') + QUOTA_LIMIT_PATH + query, key = key)
        } catch (error: Exception) {
            if (error is kotlinx.coroutines.CancellationException) throw error
            return Result.failure(error as? UsageProviderException ?: UsageProviderException.Transport())
        }

        return runCatching {
            mapResponses(modelResponse, quotaResponse, credential.id)
        }.recoverCatching { error ->
            if (error is UsageProviderException) throw error
            throw UsageProviderException.InvalidResponse()
        }
    }

    private suspend fun request(url: String, key: String): ProviderHttpResponse {
        val response = transport.execute(
            ProviderHttpRequest(
                method = "GET",
                url = url,
                headers = mapOf(
                    "Authorization" to key,
                    "Content-Type" to "application/json",
                    "Accept-Language" to "en-US,en"
                )
            )
        )
        if (response.statusCode !in 200..299) {
            throw when (response.statusCode) {
                401, 403 -> UsageProviderException.Unauthorized()
                429 -> UsageProviderException.RateLimited()
                in 500..599 -> UsageProviderException.ProviderUnavailable()
                else -> UsageProviderException.InvalidResponse()
            }
        }
        return response
    }

    private fun mapResponses(modelResponse: ProviderHttpResponse, quotaResponse: ProviderHttpResponse, credentialId: java.util.UUID): UsageSnapshot {
        val modelRoot = ProviderJson.parse(modelResponse.body.decodeToString()).asObjectOrNull()
        val quotaRoot = ProviderJson.parse(quotaResponse.body.decodeToString()).asObjectOrNull()

        val quotaMetrics = quotaMetrics(quotaRoot)
        val metrics = buildList {
            addAll(quotaMetrics.filter { it.id != "zcode-mcp" })
            modelCallMetric(modelRoot)?.let { add(it) }
            quotaMetrics.firstOrNull { it.id == "zcode-mcp" }?.let { add(it) }
        }
        if (metrics.isEmpty()) throw UsageProviderException.InvalidResponse()

        return UsageSnapshot(
            credentialId = credentialId,
            fetchedAt = clock.instant(),
            planName = "Coding Plan",
            metrics = metrics
        )
    }

    private fun quotaMetrics(root: JsonObject?): List<UsageMetric> {
        val limits = root?.objOrNull("data")?.arrayOrNull("limits")
            ?: root?.arrayOrNull("limits")
            ?: return emptyList()

        var fiveHour: UsageMetric? = null
        var weekly: UsageMetric? = null
        var zcodeMcp: UsageMetric? = null

        for (item in limits) {
            val obj = item.asObjectOrNull() ?: continue
            val percentage = obj.decimalOrNull("percentage") ?: continue
            val type = obj.stringOrNull("type") ?: "quota"
            val unit = obj.intOrNull("unit")
            val number = obj.intOrNull("number")
            val windowEnd = obj.longOrNull("nextResetTime")?.let { Instant.ofEpochMilli(it) }

            when {
                type == "TOKENS_LIMIT" && unit == 3 && number == 5 ->
                    fiveHour = percentMetric(id = "five-hour", label = "5 小时用量", percentage = percentage, windowEnd = windowEnd)
                type == "TOKENS_LIMIT" && unit == 6 && number == 1 ->
                    weekly = percentMetric(id = "weekly", label = "每周用量", percentage = percentage, windowEnd = windowEnd)
                type == "TIME_LIMIT" -> {
                    val limit = obj.decimalOrNull("usage") ?: BigDecimal(100)
                    val used = obj.decimalOrNull("currentValue") ?: percentage
                    val remaining = obj.decimalOrNull("remaining")
                        ?: limit.subtract(used).max(BigDecimal.ZERO)
                    zcodeMcp = UsageMetric(
                        id = "zcode-mcp",
                        label = "MCP 调用量",
                        used = used,
                        limit = limit,
                        remaining = remaining,
                        value = used,
                        unit = UsageMetricUnit.Request,
                        windowEnd = windowEnd,
                        presentation = UsageMetricPresentation.Value,
                        semantic = UsageMetricSemantic.UsedQuota,
                        healthState = UsageMetricHealthEvaluator.fromPercent(percentage.toDouble())
                    )
                }
            }
        }
        return listOfNotNull(fiveHour, weekly, zcodeMcp)
    }

    private fun modelCallMetric(root: JsonObject?): UsageMetric? {
        val count = root?.objOrNull("data")?.objOrNull("totalUsage")
            ?.decimalOrNull("totalModelCallCount") ?: return null
        return UsageMetric(
            id = "model-calls",
            label = "模型调用量",
            value = count,
            unit = UsageMetricUnit.Request,
            presentation = UsageMetricPresentation.Value,
            semantic = UsageMetricSemantic.Value,
            healthState = UsageMetricHealthState.Normal
        )
    }

    private fun percentMetric(id: String, label: String, percentage: BigDecimal, windowEnd: Instant?): UsageMetric =
        UsageMetric(
            id = id,
            label = label,
            used = percentage,
            limit = BigDecimal(100),
            remaining = BigDecimal(100).subtract(percentage).max(BigDecimal.ZERO),
            unit = UsageMetricUnit.Token,
            windowEnd = windowEnd,
            presentation = UsageMetricPresentation.Progress,
            semantic = UsageMetricSemantic.UsedQuota,
            healthState = UsageMetricHealthEvaluator.fromPercent(percentage.toDouble())
        )

    /** Start of the model-usage query window: one day before now, local calendar. */
    private fun windowStart(): String = format(LocalDateTime.now(clock).minusDays(1))

    /** End of the query window: the current hour with minutes/seconds set to 59, like the macOS client. */
    private fun windowEnd(): String {
        val now = LocalDateTime.now(clock)
        return format(now.withMinute(59).withSecond(59).withNano(0))
    }

    private fun format(value: LocalDateTime): String = value.format(FORMATTER)

    private fun encode(value: String): String {
        val builder = StringBuilder()
        for (byte in value.toByteArray(Charsets.UTF_8)) {
            val c = byte.toInt().toChar()
            if (c in 'A'..'Z' || c in 'a'..'z' || c in '0'..'9' || c == '-' || c == '_' || c == '.' || c == '~' || c == ':') {
                builder.append(c)
            } else {
                builder.append('%')
                builder.append("%02X".format(byte))
            }
        }
        return builder.toString()
    }

    companion object {
        const val DEFAULT_BASE_URL = "https://api.z.ai"
        private const val MODEL_USAGE_PATH = "/api/monitor/usage/model-usage"
        private const val QUOTA_LIMIT_PATH = "/api/monitor/usage/quota/limit"

        private val FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    }
}
